src/python/pants/engine/internals/graph.py

```diff
@@ -344,6 +346,11 @@ async def resolve_targets(
     return Targets(expanded_targets)
 
 
+@rule
+def filter_targets(targets: Targets, specs_filter: SpecsFilter) -> FilteredTargets:
```

Comment:
Since this filtering should only ever be applied to the (global, although that's hard to enforce) `Specs` (as implied by `SpecsFilter`), should this rule maybe be:
```suggestion
@rule
def filter_targets(specs: Specs, specs_filter: SpecsFilter) -> FilteredTargets:
    targets = await Get(Targets, Specs, specs)
```
?

The difference is that it could only be triggered for things which are actually `Specs`, rather than anything which can be converted into `Targets` (which includes literal `Addresses`, for example).

Reply:
That might be a good safety hatch. We can loosen if the need arises. I can try that and see if the rule graph compiles

Reply:
Eh, I couldn't get it to compile and it causes complications inside our `Owners` rule, which gives `AddressSpecs` rather than `Specs`